image.h

```diff
@@ -1,20 +1,42 @@
 #pragma once
 #include <cstdint>
 #include <iostream>
+#include <functional>
+
+const int PixelBrightenerFactor = 25;
 
 class Image {
 public:
-	Image(uint16_t rows, uint16_t columns): m_rows(rows), m_columns(columns) {
-		pixels = new uint8_t[m_rows * m_columns];
-	}
-	~Image() {
-		std::cout << "freeing pixel memory.\n";
-		delete[] pixels;
-	}
-	bool imageSizeIsValid() {
-		return m_rows <= 1024 && m_columns <= 1024;
-	}
-	const uint16_t m_rows;
-	const uint16_t m_columns;
-	uint8_t* pixels; // max 1k x 1k image
+    using PixelProcessCallback = std::function < uint8_t(uint8_t, uint8_t)>;
+
+    Image(uint16_t rows, uint16_t columns) : m_rows(rows), m_columns(columns) {
+        m_pixels = new uint8_t[m_rows * m_columns];
+    }
+
+    ~Image() {
+        std::cout << "freeing pixel memory.\n";
+        delete[] m_pixels;
+    }
+
+    bool isSizeValid() {
+        return m_rows <= 1024 && m_columns <= 1024;
+    }
+
+    void pixelRunner(PixelProcessCallback pixelProcessor, std::shared_ptr<Image> imageToAdd = nullptr) {
+        for (int x = 0; x < m_rows; x++) {
+            for (int y = 0; y < m_columns; y++) {
+                int pixelIndex = x * m_columns + y;
+                if (imageToAdd) {
```

Comment:
This `if-else` construct will keep getting more complex, as we add more types of processing. Anyway `pixelProcessor` is a callback which can process pixels any way it wants. Why not pack the variations into different callbacks?
Like have one function to add the `PixelBrightenerFactor`, another to use the `imageToAdd` to add up the respective pixels, etc?

Reply:
Done